test_pyproject_save_files.py

```
@@ -0,0 +1,198 @@
+
+
+def test_parse_record_tensorflow():
+    long = "tensorflow_core/include/tensorflow/core/common_runtime/base_collective_executor.h"
+    record_path = SITEARCH / "tensorflow-2.1.0.dist-info/RECORD"
+    record_content = [
+        ["../../../bin/toco_from_protos", "sha256=hello", "289"],
+        [f"../../../lib/python3.7/site-packages/{long}", "sha256=darkness", "1024"],
+        ["tensorflow-2.1.0.dist-info/METADATA", "sha256=friend", "2859"],
+    ]
+    output = list(parse_record(record_path, record_content))
+    pprint(output)
+    expected = [
+        BINDIR / "toco_from_protos",
+        SITELIB / long,
+        SITEARCH / "tensorflow-2.1.0.dist-info/METADATA",
+    ]
+    assert output == expected
+
+
+# packagename: expected path in buildroot
+TEST_RECORDS = {
+    "kerberos": SITEARCH / "kerberos-1.3.0.dist-info/RECORD",
+    "requests": SITELIB / "requests-2.22.0.dist-info/RECORD",
+    "tensorflow": SITEARCH / "tensorflow-2.1.0.dist-info/RECORD",
+    "tldr": SITELIB / "tldr-0.5.dist-info/RECORD",
+    "mistune": SITEARCH / "mistune-0.8.3.dist-info/RECORD",
+}
+
+
+def remove_executables(expected):
+    return [p for p in expected if not p.startswith(str(BINDIR))]
+
+
+@pytest.mark.parametrize("include_executables", (True, False))
+@pytest.mark.parametrize("package, glob, expected", EXPECTED_FILES)
+def test_generate_file_list(package, glob, expected, include_executables):
+    paths_dict = EXPECTED_DICT[package]
+    modules_glob = (glob,)
+    if not include_executables:
+        expected = remove_executables(expected)
+    tested = generate_file_list(paths_dict, modules_glob, include_executables)
+
+    assert tested == expected
+
+
+@pytest.mark.parametrize(
+    "arguments, output",
+    [
+        (["requests*", "kerberos", "+bindir"], (["requests*", "kerberos"], True)),
+        (["tldr", "tensorf*"], (["tldr", "tensorf*"], False)),
+        (["+bindir"], ([], True)),
+        (["+kinkdir"], (["+kinkdir"], False)),
```

These should be toplevel module names; should we ban `+kinkdir` with an `isidentifier` check?

That would also ban subpackages of namespaces, like if someone wanted to package `sphinxcontrib.mycoolplugin` separately. That might be a valid use case, but we're not prepared for it so better fail early.

---

using isidentifier with a glob will be tricky thou

---

I've added this as a goal to the TODO above. However I'm still not sure how to validate that a glob can represent a valid identifier. We could:

 - convert `*` and `?` to a single valid character that can be used anywhere (such as `x`)
 - expand `[seq]` to several identifiers (e.g. `s`, `e` and `q`) and check all (hoping the sequence  is not thousands charters long)
 - expand `[!seq]` to a single valid character (hoping the sequence doesn't contain all valid characters)

However this seems very fragile. I don't want to maintain and test that.

What if we only explicitly check for cases we care about:

 - treat everything that starts with `+` as an option and error on invalid options (that also lets us add more options later)
 - provide a nice error if there is a dot
 - let all other cases fail later, when the glob matches no module

---

Right! When reviewing that line I missed that these are globs, not package names.
The second option looks good!

---

Thanks. All part of the TODO (some of it already implemented). Marking as resolved.